src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,65 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  constexpr int FACTOR_INCREASE = 3; //approximate value by which total size of chain, tag and time increases after encoding into bufferlist
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->rgw_max_chunk_size) {
+    cls_rgw_obj_chain broken_chain;
+    bool chain_sent = false;
+    auto tag_size = tag.length(), expiration_time_size = sizeof(ceph::real_time);
+    auto adjusted_max_write_size = ((cct->_conf->rgw_max_chunk_size)/FACTOR_INCREASE); //approx upper limit
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - rgw_max_chunk_size is: " << cct->_conf->rgw_max_chunk_size << dendl;
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - adjusted_max_write_size is: " << adjusted_max_write_size << dendl;
+    auto cumulative_size = 0;
+    for (auto it = chain.objs.begin(); it != chain.objs.end(); it++) {
+      auto size = it->pool.size() + it->loc.size() + it->key.name.size() + it->key.instance.size();
+      cumulative_size += size;
+      auto total_size = cumulative_size + tag_size + expiration_time_size;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - cumulative_size is: " << cumulative_size << dendl;
+      ldpp_dout(this, 20) << "RGWGC::send_split_chain - total_size is: " << total_size << dendl;
+
+      if (total_size < adjusted_max_write_size) { //continue adding to chain
+        broken_chain.objs.emplace_back(*it);
+        chain_sent = false;
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - less than, add to broken chain and continue" << dendl;
+        continue;
+      }
+      if (total_size == adjusted_max_write_size) { //add to chain and send to gc
+        broken_chain.objs.emplace_back(*it);
+        ldpp_dout(this, 20) << "RGWGC::send_split_chain - equal to, add to broken chain and send chain" << dendl;
+        auto ret =  send_split_chain(broken_chain, tag);
```

Comment:
let's try to avoid recursion here

Reply:
this is bug, must be send_chain(), I will correct it